### PR TITLE
UAVCAN/UART1 Moving Baseline for U-Blox F9P

### PR DESCRIPTION
**Pull request by AlexKlimaj**
This PR adds moving baseline functionality on the ARK RTK GPS using the U-Blox F9P.

PXL_20211011_203619308

### CODE REVIEW

File: src/ubx.cpp [hunk 1]
```diff
@@ -107,11 +107,15 @@ GPSDriverUBX::configure(unsigned &baudrate, const GPSConfig &config)
 	if (_interface == Interface::UART) {
 
 		/* try different baudrates */
-		const unsigned baudrates[] = {38400, 57600, 9600, 115200, 230400};
+		const unsigned baudrates[] = {38400, 57600, 9600, 115200, 230400, 460800, 921600};
 
 		unsigned baud_i;
 		unsigned desired_baudrate = auto_baudrate ? UBX_BAUDRATE_M8_AND_NEWER : baudrate;
 
+		if ((_mode == UBXMode::RoverWithMovingBaseUART1) || (_mode == UBXMode::MovingBaseUART1)) {
```
Comment (bkueng) [inline src/ubx.cpp:115]: Can you check for `auto_baudrate` as well?

File: src/ubx.cpp [hunk 2]
```diff
@@ -713,6 +726,106 @@ int GPSDriverUBX::configureDevice(const GNSSSystemsMask &gnssSystems)
...
+		// cfgValset<uint8_t>(UBX_CFG_KEY_MSGOUT_UBX_NAV_RELPOSNED_UART2, 0, cfg_valset_msg_size);
+
+		// cfgValset<uint8_t>(UBX_CFG_KEY_MSGOUT_RTCM_3X_TYPE4072_0_UART1, 0, cfg_valset_msg_size);
+		// cfgValset<uint8_t>(UBX_CFG_KEY_MSGOUT_RTCM_3X_TYPE4072_1_UART1, 0, cfg_valset_msg_size);
+		// cfgValset<uint8_t>(UBX_CFG_KEY_MSGOUT_RTCM_3X_TYPE1077_UART1, 0, cfg_valset_msg_size);
+		// cfgValset<uint8_t>(UBX_CFG_KEY_MSGOUT_RTCM_3X_TYPE1087_UART1, 0, cfg_valset_msg_size);
+		// cfgValset<uint8_t>(UBX_CFG_KEY_MSGOUT_RTCM_3X_TYPE1097_UART1, 0, cfg_valset_msg_size);
+		// cfgValset<uint8_t>(UBX_CFG_KEY_MSGOUT_RTCM_3X_TYPE1127_UART1, 0, cfg_valset_msg_size);
+		// cfgValset<uint8_t>(UBX_CFG_KEY_MSGOUT_RTCM_3X_TYPE1230_UART1, 0, cfg_valset_msg_size);
+		// cfgValset<uint8_t>(UBX_CFG_KEY_MSGOUT_RTCM_3X_TYPE1074_UART1, 0, cfg_valset_msg_size);
```
Comment (bkueng) [inline src/ubx.cpp:751]: Can you remove the commented ones?

File: src/ubx.cpp [hunk 3]
```diff
@@ -713,6 +726,106 @@ int GPSDriverUBX::configureDevice(const GNSSSystemsMask &gnssSystems)
...
+
+		if (!sendMessage(UBX_MSG_CFG_VALSET, (uint8_t *)&_buf, cfg_valset_msg_size)) {
+			return -1;
+		}
+
+		if (waitForAck(UBX_MSG_CFG_VALSET, UBX_CONFIG_TIMEOUT, true) < 0) {
+			return -1;
+		}
+
+		setBaudrate(uart1_baudrate);
```
Comment (bkueng) [inline src/ubx.cpp:778]: Why do you set the baudrate here again?

Comment (AlexKlimaj) [reply src/ubx.cpp:778]: This changes the baudrate in PX4 to match what it gets set to on the F9P. Without it, PX4 loses communication and needs to re-auto baud.

File: src/ubx.cpp (hunk 1 above)
Comment (AlexKlimaj) [reply src/ubx.cpp:115]: I'm not sure I understand what that accomplishes?

File: src/ubx.cpp (hunk 3 above)
Comment (bkueng) [reply src/ubx.cpp:750]: But the baudrate for uart1 is already set above, so `cfgValset<uint32_t>(UBX_CFG_KEY_CFG_UART1_BAUDRATE, uart1_baudrate, cfg_valset_msg_size);` is not needed here. Am I missing something?

Comment (AlexKlimaj) [reply src/ubx.cpp:750]: `cfgValset<uint32_t>(UBX_CFG_KEY_CFG_UART1_BAUDRATE, uart1_baudrate, cfg_valset_msg_size);` sets the baudrate on the F9P side. `setBaudrate(uart1_baudrate);` sets it on the PX4 side to match.

**Review by dagar — APPROVED** [review]

File: src/ubx.cpp (hunk 3 above)
Comment (bkueng) [reply src/ubx.cpp:750]: @AlexKlimaj could you please address this?

Comment (AlexKlimaj) [reply src/ubx.cpp:750]: To do get rtcm and position data over the single uart1, it needs to be set to 921600 in that case. If I moved the set baudrate to where you suggest, I'd have to add another check for what UBXMode is set to.

**Pull request merged**